### PR TITLE
feat: add JS config handling to webpack and jest configs

### PR DESCRIPTION
**Pull request by jsnwesson**
This PR includes the following changes: 
- JS-based configuration is made available to Webpack Prod by passing in an environment variable
- Webpack Dev config can now use the `PORT` assigned in `env.config.js`. Otherwise checks `process.env.PORT` or falls back to 8080.
- Jest config now searches for either `env.config.js` or `env.config.jsx` for environment variables. 
- updated README to provide steps for using JS config in production and testing. 

Resolves issues #513 and #514

### CODE REVIEW

File: config/webpack.prod.config.js [hunk 1]
```diff
@@ -45,12 +65,12 @@ if (process.env.ENABLE_NEW_RELIC !== 'false') {
     agentID: process.env.NEW_RELIC_AGENT_ID || 'undefined_agent_id',
     trustKey: process.env.NEW_RELIC_TRUST_KEY || 'undefined_trust_key',
     licenseKey: process.env.NEW_RELIC_LICENSE_KEY || 'undefined_license_key',
-    applicationID: process.env.NEW_RELIC_APP_ID || 'undefined_application_id',
+    applicationID: envConfig.NEW_RELIC_APP_ID || process.env.NEW_RELIC_APP_ID || 'undefined_application_id',
```
Comment (adamstankiewicz) [inline config/webpack.prod.config.js:68]: [curious] Would we expect the other `process.env.NEW_RELIC_*` env vars might be usable with `envConfig` as well (e.g., `agentID`, `trustKey`, `licenseKey`, etc.)?

File: README.md [hunk 2]
```diff
@@ -218,6 +217,31 @@ locally. To serve a production build locally:
...
+
+The only requirement is that the filepath end with `env.config.*`, where either `.js` or `.jsx` as the extension
+
+    // examples of acceptable filepaths
+
+    JS_CONFIG_FILEPATH="{HOME}/frontends/frontend-app-learner-dashboard/prod.env.config.js"
+
+    JS_CONFIG_FILEPATH="{HOME}/frontends/frontend-app-profile/stage.env.config.jsx"
+
+## Requiring Jest to reference env.config.js
```
[the remark this answers is not in the document]
Comment (jsnwesson) [reply README.md:233]: Actually, `env.config.js` isn't required to run tests, but if there are any env variables or configuration in the env.config.js file that should be used for tests, this is what will be needed in the `setupTest.js` file.

File: config/webpack.prod.config.js (hunk 1 above)
Comment (jsnwesson) [reply config/webpack.prod.config.js]: I actually went ahead and removed these changes as this would've meant using a `require` statement in this config to fetch the environment variables, and that got in the way of using `import` in the `env.config.js` file. 

This will mean that there's no clear solution yet for if we ever want to replace the `process.env.ENV_VAR`s in this file or `webpack.dev.config.js`.

File: config/jest.config.js [hunk 3]
```diff
@@ -3,11 +3,15 @@ const fs = require('fs');
 
 const presets = require('../lib/presets');
 
+// This assigns the envConfigPath filepath based on whether env.config exists, otherwise it uses the fallback filepath.
 let envConfigPath = path.resolve(__dirname, './jest/fallback.env.config.js');
-const appEnvConfigPath = path.resolve(process.cwd(), './env.config.js');
+const appEnvConfigPathJs = path.resolve(process.cwd(), './env.config.js');
+const appEnvConfigPathJsx = path.resolve(process.cwd(), './env.config.jsx');
```
[the remark this answers is not in the document]
Comment (bradenmacdonald) [reply config/jest.config.js:10]: I would! ;)

File: config/jest.config.js [hunk 4]
```diff
@@ -3,11 +3,16 @@ const fs = require('fs');
...
 
-if (fs.existsSync(appEnvConfigPath)) {
-  envConfigPath = appEnvConfigPath;
+if (fs.existsSync(appEnvConfigPathJs)) {
+  envConfigPath = appEnvConfigPathJs;
+} else if (fs.existsSync(appEnvConfigPathJsx)) {
+  envConfigPath = appEnvConfigPathJsx;
```
Comment (brian-smith-tcril) [inline config/jest.config.js:15]: Is there a specific reason this logic prioritizes `.js` over `.jsx`? I don't have strong feelings either way, but the decision around the prioritization should be documented.

File: config/jest.config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (leangseu-edx) [reply config/jest.config.js:10]: can we use https://www.npmjs.com/package/glob to find the file with regex instead? It is easier to improve in the future.

**Comment by MaxFrank13** [comment]
Hello all! I have taken over this PR as Jason is out for the next couple weeks. I have read through comments/code and have some questions to help get me up to speed.

1. `.ts(x)` support and prioritization
   - The recommendation to use `glob` seems straightforward but would introduce another dependency for `frontend-build` — are there any objections to adding this dependency?
   - W.r.t prioritization, I'm having a hard time understanding why there would ever be more than one configuration file. That sounds like an implementation mistake seeing as we don’t have a mechanism for combining multiple config files in the first place. So anyone with two config files has presumably made an error in their setup, no? Regardless, I'll be happy to sort this out... once I'm less confused 😆 
   
2. Same changes in `jest` config to go in `dev` and `dev-stage` configs
   - Should we add [the config changes](https://github.com/openedx/frontend-build/blob/9acebc63c5a267f96a7d7f45eca9e0d81f103a0d/config/jest.config.js#L9-L16) to the other config files? If my understanding is correct, this would enable us to use the config file path rather than putting env values directly in the MFE’s `env.config.{js/jsx/ts/tsx}` file
   - From what I gathered in the discussion, this was recommended as well as extracting the functionality into a utility to reduce code duplication. Just making sure I understand the suggestion before doing any work.
 
3. Need for specific community use case
   - Is this still true post-adoption of FPF? 
   - In other words, is FPF a sufficient primary use case?
   
4. Clarification around `jest.config` and why it’s been modified
   - Allows tests to access `env.config` values if needed — do we need specific examples of when this would be done?
   - Do we have examples?

cc: @adamstankiewicz @arbrandes @jsnwesson @brian-smith-tcril @leangseu-edx

**Comment by arbrandes** [comment]
As discussed in [the last FWG meeting](https://openedx.atlassian.net/wiki/spaces/COMM/pages/4194893825/2024-04-25+Frontend+Working+Group+Meeting+Notes), I don't see the point of adding an `.env` variable, here.  We already have [webpack configuration](https://github.com/openedx/frontend-build/blob/master/config/webpack.common.config.js#L13) that pulls in `env.config.js*` if present.  It shouldn't be frontend-build's responsibility to copy that file in.

**Comment by MaxFrank13** [comment]
Closed as we aren't going forward with changes to the `prod.config` and don't see an immediate need for the `jest.config` changes. It may just need to confusion as we haven't fully migrated to JS configs yet.
